docs/SequenceType.md

```diff
@@ -0,0 +1,318 @@
+could be marked as `escaped` and then no deallocation would be added by the 
+deallocation pass. Such change will involve how to add clone op with the present of 
+control flow. Our current solution is based the existing deallocation pass.
+
+### Store an element into a sequence
+To avoid the deallocation of the element, we can save a copy of the element into the
+sequence. To get a copy, we need to allocate a memref and then use memref.copy
+to copy the value. However, deallocation pass will again add deallocation to
+the newly allocated memref. To avoid this issue, we should wrap all the operations of
+memref the allocation, copy and store into one krnl Op, krnlSeqStoreOp, which will be
+lowered AFTER the deallocation pass.
+
+Since the type of sequence element is a super type for all possible element, memref.cast may be
+needed before the store. Here is a code example for KrnlSeqStoreOp:
+```
+// %seq : memref<?xmemref<?x2xf32>>
+// %element : memref<3x2xf32>
+// %pos : Index
+      %33 = memref.alloc(%32) {alignment = 16 : i64} : memref<3x2xf32>
+      memref.copy %element, %33 : memref<3x2xf32> to memref<3x2xf32>
+      %34 = memref.cast %33 : memref<3x2xf32> to memref<?x2xf32>
+      memref.store %34, %seq[%pos] : memref<?xmemref<?x2xf32>>
+```
+
+### Allocate a sequence
+Though the basic operation of sequence allocation is just memref.alloc, we introduced
+KrnlSeqAllocOp so that we can define a customized deallocation function for sequence.
+We use interface in MLIR Bufferization to specify that KrnlSeqAllocOp has allocation 
+traits and a customized free function, which will perform a deep deallocation for the
+elements as well as the sequence itself. Currently, the KrnlSeqDeallocOp is used for the
+deallocation and it will be lowered to scf and memref after deallocation pass.
+
+### Load an element from a sequence
+A memref.load could be used to load an element from a sequence. But the loaded memref may
+have a life span longer than the sequence itself, and will have dangling data pointer after
+the sequence has been freed.
+
+To overcome this issue, KrnlSeqExtractOp is introduced. This Op will use memref.load to
+load the element, then use allocate a new memref and copy the data, and finally return
+the copied memref.  This Op is marked with allocation interface and the deallocation pass will insert deallocation for the returned memref automatically. 
+
+### Construct a new sequence from an old sequence
+The sequence ops, SequenceInsert and SequenceErase, construct a new sequence
+with the elements from an old sequence. It is fine to use KrnlSeqExtractOp to get an
+element from the old sequence and use KrnlSeqStoreOp to store it into the new sequence.
+But there are two copy operations for the element. Since it is known that the loaded
+element is only used here while the old sequence is still live, a regular memref.load 
+can be used to avoid one copy. This is a simple optimization for sequence lowering.
+
+## Example
```

Comment:
Would it make sense to introduce each of the seq op and describe their inputs... I find it a bit harder to follow the examples without knowing the semantic, at a high level.

Also, could you describe at a high level what the function does.... "insert addition to an input sequence at position X...."